Python client implementation

jemorriso opened the pull request:
This is not yet completed - opening the pull request to get feedback.

The API client is implemented in class `Rundown` in `rundown.py`. Each method implements an API route, and returns a nested object based on the JSON sent back from the route. The object classes are found in the `resources` folder.

The user can pass in their preferred timezone into `Rundown` and all responses will have dates in that timezone.

Features:
- handles both  RapidAPI (tested) and The Rundown API (untested) authentication
- automatic timezone handling
- [automated test suite](https://app.circleci.com/settings/project/github/JeMorriso/therundown-python) hosted on CircleCI

Not implemented:
- automatic odds type handling: [Pydantic](https://pydantic-docs.helpmanual.io/) is being used for JSON validation / parsing which is hugely helpful for nested objects, but has some limitations. It's definitely possible to handle different odds types (eg Decimal) but it would just require a workaround. If this is a desired feature doing conversions on the server side is an alternative.
 - `update` method  on `resources.Events`: my idea was to have a method that uses `delta_last_id` and merges updated events with unchanged ones. This would be implemented as a convenient alternative to using `Rundown.events_delta`.
 - other higher level features: I wanted to keep it simple for now.

To do:
- testing edge cases
- packaging and release
- README / documentation

You can experiment with the code by installing the `rundown` folder as an editable dependency and making calls to the API through the `Rundown` class.

jemorriso committed:
Initial commit

jemorriso committed:
Add CircleCI configuration

jemorriso committed:
Add build and test workflow

jemorriso committed:
Add arrow, pydantic

arrow for datetime manipulation, pydantic for data validation. Also add
to dev-dependencies python-dotenv for environment variables and
ipykernel for Jupyter lab support of poetry environment.

jemorriso committed:
Autoload .env variables using python-dotenv

jemorriso committed:
Add Rundown constructor and private helper classes

Store user configuration as attributes on Rundown. Authorization classes
add necessary configuration to access either RapidAPI or The Rundown's
API.

jemorriso committed:
Add test for Auth factory

jemorriso committed:
Add internal methods for making get request

jemorriso committed:
Add pytest-vcr, monkey patch Rundown class

Save request JSON body along with VCR cassettes.

jemorriso committed:
Test internal methods supporting GET requests

jemorriso committed:
Add VCR cassettes and JSON for new tests

jemorriso committed:
Make sure no API calls are made from CircleCI

jemorriso committed:
Add, test utility function for getting UTC offset

jemorriso committed:
Fix JSON file names

Add build_file_name function to match VCR cassette file names.

jemorriso committed:
Add and test Rundown.sports method

jemorriso committed:
Add and test Rundown.dates_by_sport method

jemorriso committed:
Add and test Rundown.sportsbook method

jemorriso committed:
Add and test Rundown.teams_by_sport method

jemorriso committed:
Add UTC shift function

jemorriso committed:
Use utc_shift instead of utc_offset

jemorriso committed:
Add cassettes and JSON with correct UTC shift

jemorriso committed:
Add events by date methods

jemorriso committed:
Add tests, cassettes for events by date methods

jemorriso committed:
Fix typo on include argument

jemorriso committed:
Fix type of include argument

jemorriso committed:
Add new test cases

jemorriso committed:
Add money, spread and total methods

jemorriso committed:
Add tests for moneyline, spread, total methods

jemorriso committed:
Add event and events_delta methods

jemorriso committed:
Test event and events_delta

jemorriso committed:
Add and test Rundown.schedule_by_sport

jemorriso committed:
Add Models

Improperly commented for now.

jemorriso committed:
Return appropriate resource types from methods

jemorriso committed:
Handle new return types from Rundown methods

jemorriso committed:
Add user context using context variables

Solution to passing dynamic context into validators suggested by Pydantic author
in this (github issue)[https://github.com/samuelcolvin/pydantic/issues/1170].
Kind of hacky but context manager allows code to be well isolated.

jemorriso committed:
Add validators file for shared validators

jemorriso committed:
Remove odds_type configuration

Remove odds_type validator, context_odds_type context variable, and
odds_type instance attribute from Rundown.

jemorriso committed:
Add timezone validators to resources

jemorriso committed:
Fix type hints and comments

jemorriso committed:
Delete old docstrings

jemorriso committed:
Simplify module, remove unnecessary functions

jemorriso committed:
Update tests

jemorriso committed:
Add timezone details in Rundown docstring

fjordan commented on rundown/resources/event.py:
```diff
@@ -0,0 +1,75 @@
+    event_date: str
+    rotation_number_away: int
+    rotation_number_home: int
+    score: Optional[Score]  # not set for future events
```
We generally try to have games populated with some score information (venue, description, broadcast, etc.) around 100 days or so before the event starts, but there is a possibility it will not be available.

jemorriso replied:
Good to know. I will change the comment, but I can leave it as Optional if there is a chance that it won't exist.

jemorriso replied:
Changed comment.

fjordan commented on rundown/resources/date.py:
```diff
@@ -0,0 +1,27 @@
+        correct with respect to timezone. So it is only necessary to truncate the date
+        string.
+        """
+        naive_dt_str = v.split("+")[0]
```
Can python not natively parse ISO8601? It looks like with `arrow` you can simply use the `.get(<iso8601-str>)` string and go from there.

jemorriso replied:
It can. I noticed a bug in `/sports/<sport-id>/dates` route where when you change the offset the timezone doesn't change, but the time does. E.G:

```python
r = Rundown(<api-key>, timezone='America/Vancouver')
s = Rundown(<api-key>, timezone='UTC')
rd = r.dates_by_sport(6)
sd = s.dates_by_sport(6)

print(rd)
# output
{'dates': ['2021-04-05T16:00:00+0000',
  '2021-04-06T16:00:00+0000',
  '2021-04-07T14:00:00+0000',
  '2021-05-05T16:00:00+0000']}

print(sd)
# output
{'dates': ['2021-04-05T23:00:00+0000',
  '2021-04-06T00:00:00+0000',
  '2021-04-07T00:00:00+0000',
  '2021-04-08T01:00:00+0000',
  '2021-05-05T23:00:00+0000']}
```

So I handled this by truncating off the incorrect timezone part.

jemorriso replied:
Ideally we could fix it on the server side and then this code wouldn't be necessary

fjordan replied:
>  I noticed a bug in /sports/<sport-id>/dates route where when you change the offset the timezone doesn't change, but the time does

Yeah, this endpoint is confusing. It is intentionally always returning the times of the 24-hour windows in UTC instead of the localized timezone, although now that you bring this up it make more sense to do what you're suggesting.

fjordan commented on rundown/rundown.py:
```diff
@@ -0,0 +1,433 @@
+        self.api_url = "https://therundown-therundown-v1.p.rapidapi.com"
+        self.headers = {
+            "x-rapidapi-key": self.api_key,
+            "x-rapidapi-host": self.api_host,
```
I couldn't find where this is referenced or used from the RapidAPI docs. What's the purpose of this header?

jemorriso replied:
This is from the endpoints documentation at https://rapidapi.com/therundown/api/therundown. Documentation specifies that 'x-rapidapi-key' and 'x-rapidapi-host' are required header parameters.

fjordan commented on rundown/rundown.py:
```diff
@@ -0,0 +1,433 @@
+
+    def __init__(self, api_key: str):
+        self.api_key = api_key
+        self.api_host = "therundown-therundown-v1.p.rapidapi.com"
```
If this is containing the host and URL in addition the keys, should it be renamed to something like `_RadpiAPI` or `_RapidAPIBase` or similar instead? 

Also, it may be a good idea to add a version param here as well.

jemorriso replied:
Changed to `_RundownBase` and `_RapidAPIBase`. I don't see anything about versions in the documentation, would this just be handling future version upgrades? I could add a parameter to `Rundown` which allows the user to specify which API version they are using, e.g. 'v2', which would get passed to the base classes and the configuration would be changed accordingly.

fjordan replied:
>  I don't see anything about versions in the documentation, would this just be handling future version upgrades?

From what I can tell, it is actually the `x-rapidapi-host` header that may be used to specify the version given the current contains the word `v1`, although I could be wrong. Because of that, we probably don't need a version, or, we could try to map the version in the configuration to the `api_host` value. 

What do you think? I wish this were more clear on RapidAPI's part.

fjordan commented on rundown/rundown.py:
```diff
@@ -0,0 +1,433 @@
+            (case insensitive).
+        timezone: Your preferred timezone.
+            The following formats are accepted:
+            - A str describing a timezone, similar to ‘US/Pacific’, or ‘Europe/Berlin’.
+            - A str in ISO 8601 style, as in ‘+07:00’.
+            - A str, one of the following: ‘local’, ‘utc’, ‘UTC’.
```
👍

fjordan commented on rundown/resources/line.py:
```diff
@@ -0,0 +1,92 @@
+    affiliate_id: int
+
+
+class Moneyline(Line):
+    moneyline_away: Optional[Union[int, float]] = Field(...)
+    moneyline_away_delta: Optional[int] = Field(...)
+    moneyline_home: Optional[int] = Field(...)
+    moneyline_home_delta: Optional[int] = Field(...)
+    moneyline_draw: Optional[int] = Field(...)
+    moneyline_draw_delta: Optional[int] = Field(...)
+
+
+class MoneylinePeriod(Moneyline):
+    period_id: int
+    period_description: str
```
nicely abstracted 👌

jemorriso replied:
Thanks!

fjordan commented on rundown/rundown.py:
```diff
@@ -0,0 +1,433 @@
+
+    def __init__(self, api_key: str):
+        self.api_key = api_key
+        self.api_url = "https://therundown.io/api/v1"
```
We should extract the version here into a separate var

fjordan commented on rundown/rundown.py:
```diff
@@ -0,0 +1,433 @@
+    def _get(self, url: str, **params: Union[str, int, List[str]]) -> requests.Response:
+        """Make get request."""
+        res = self._session.get(url, params=params)
+        # TODO: handle 404 not found - should never happen if called through methods.
```
Depending on the timing of subsequent `/delta` requests and how many events have updated in between, it's also possible to get a 400. As a result, we should handle that error message here and send the error back to the user/logs.

fjordan commented on rundown/resources/team.py:
```diff
@@ -0,0 +1,23 @@
+    is_home: bool
+
+
+class TeamDeprecated(BaseModel):
```
👍  to naming this `TeamDeprecated`

fjordan commented on rundown/resources/team.py:
```diff
@@ -0,0 +1,23 @@
+    abbreviation: str
+
+
+class TeamNormalized(Team):
```
Can we just move all of these to a `Team` model (but using the `teams_normalized` attribute of the response?)

jemorriso replied:
Yeah that makes sense. I will do that

jemorriso replied:
I realized that the reason I broke team into `Team` and `TeamNormalized` is that `GET /sports/<sport-id>/teams` only has the data in `Team`, but all the events routes also include the data in `TeamNormalized`.

fjordan replied:
Ok, that makes sense and works for me.

fjordan commented on rundown/resources/schedule.py:
```diff
@@ -0,0 +1,35 @@
+from rundown.resources.validators import change_timezone
+
+
+class BaseSchedule(BaseModel):
```
what's the purpose of breaking it into a `BaseSchedule` and `Schedule` class?

jemorriso replied:
I noticed that `GET /sports/<sport-id>/events/<date>` returns a subset of schedule data compared to `GET /sports/<sport-id>/schedule`, so I separated it into `BaseSchedule` and `Schedule`.

Event route example return: 
```
"schedule": {
  "season_type": "Regular Season",
  "season_year": 2021,
  "event_name": "Detroit at Tampa Bay - 2021-04-03",
  "attendance": "3800"
},
```

Schedule route example return:
```
{
  "id": "91570",
  "event_uuid": "11eb-967b-cb3f5000-80e8-e33c8bac5ab1",
  "event_id": "cdccf7984487550a49d3ba746f651603",
  "sport_id": 6,
  "season_type": "Regular Season",
  "season_year": 2021,
  "away_team_id": 121,
  "home_team_id": 120,
  "away_team": "Arizona",
  "home_team": "Los Angeles",
  "date_event": "2021-04-06T02:00:00Z",
  "neutral_site": false,
  "conference_competition": false,
  "away_score": 0,
  "home_score": 0,
  "league_name": "National Hockey League",
  "event_name": "Arizona at Los Angeles - 2021-04-06",
  "event_location": "STAPLES Center",
  "attendance": 0,
  "updated_at": "2021-04-05T06:11:05.328635Z",
  "event_status": "STATUS_SCHEDULED",
  "event_status_detail": "4/5 - 10:00 PM EDT"
}
```

fjordan commented on rundown/resources/sportsbook.py:
```diff
@@ -0,0 +1,7 @@
+
+
+class Sportsbook(BaseModel):
+    affiliate_name: str
```
it would be great to provide a way to extract specific sportsbooks by name instead of just the `affiliate_id`. For example, `sportsbooks.pinnacle` or similar.

jemorriso replied:
What would be the use case of doing this? I'm thinking it could be useful to allow the user to set their preferred sportsbooks (like timezone) and then filter out results.

fjordan replied:
> What would be the use case of doing this?

It's to remember the names of the sportsbooks rather than their identifier

fjordan commented on tests/test_rundown.py:
```diff
@@ -0,0 +1,236 @@
+
+
+class TestRundown:
+    # def test bad_route(self):
+    # moneyline E.G
+    # bad event id
+    # over 500 scheduled events
+    # pass
```
this is just leftover?

jemorriso replied:
Yes. That should be removed.

jemorriso replied:
Removed.

fjordan commented on rundown/rundown.py:
```diff
@@ -0,0 +1,433 @@
+
+    def _get_events(
+        self,
+        sport_id: int,
```
it would be great to be able to support a `sport_name` param here as well (ie: `nfl` or `nhl`)

fjordan reviewed (commented):
Overall this looks amazing 👏 

I left some comments and questions below.

jemorriso committed:
Small changes based on pull request comments

Update comment in BaseEvent model. Remove commented out old code in
TestRundown.

jemorriso committed:
Update Rundown config class names

Change from using 'Auth' as a suffix to using 'Base' for _RundownBase
and _RapidAPIBase classes.

jemorriso committed:
Add static sports module

Allows for use of sport names in Rundown methods in addition to sport
ids.

jemorriso committed:
Use dict, list instead of Typing.Dict, Typing.List

Typing.Dict and Typing.List are deprecated as of Python 3.9.

jemorriso committed:
Use keyword-only arguments with *include

This is necessary because in the case that optional default arguments
are not specified, 'scores' or 'all_periods' will not be captured as
part of include list, they will be assigned as a positional argument to
the previous unassigned argument with a higher precedence.

jemorriso committed:
Use typing.Literal instead of str when possible

jemorriso committed:
Change 'date_' parameters to 'date'

jemorriso committed:
Rename 'schedule_by_sport' method to 'schedule'

jemorriso committed:
Add API test class

Add test cases in order to formalize distinction between API and client
library behaviour.

jemorriso committed:
Add Rundown._json

Allow user to access raw JSON from API.

jemorriso committed:
Fix Rundown._clean_params method

Make sure it only catches `None` and empty list, and not int = 0 for
example.

jemorriso committed:
Add utc_shift_to_tz function

Allows user to override timezone with a UTC offset, preserving API
behaviour.

jemorriso committed:
Use Arrow to simplify make_timezone validator

jemorriso committed:
Add test for Rundown._validate_sport

jemorriso committed:
Improve tests for Rundown.dates_by_sport

jemorriso committed:
Add utc_shift_to_tz to dates_by_sport

Fix offset parameter behaviour.

jemorriso committed:
Add API test of epoch response

Show that the timestamps are shifted depending on the the timezone.

jemorriso committed:
Simplify Rundown method names

'dates', 'teams', 'schedule' are implied to be by sport by the 1st
parameter, so remove 'by_sport' suffix from methods.

jemorriso committed:
Update tests with new method names

jemorriso committed:
Add _with_timezone_context decorator

Keeps code DRY.

jemorriso committed:
Update tests with keyword-only offset parameter

jemorriso committed:
Merge Event and EventLinePeriods

Use typing.Union to consolidate the two in Events resource.

jemorriso committed:
Add test showing events routes default is 'scores'

jemorriso committed:
Add test showing bad date returns today's events

jemorriso committed:
Add forgotten cassettes + JSON

jemorriso committed:
Add deepdiff to dev dependencies

jemorriso committed:
Add test for 'all_periods' and 'scores'

Result is the same with or without 'scores' if 'all_periods' is
included.

jemorriso committed:
Add test showing bad date returns todays events

jemorriso committed:
Add test showing CFL not implemented

jemorriso committed:
Use _with_timezone_context decorator with methods

jemorriso committed:
Update Event with correct typing on fields

jemorriso committed:
Update tests for events, opening, closing line

Test cases cover all sports.

jemorriso committed:
Update tests for sync with CircleCI

jemorriso committed:
Add lines tests for 'all_periods' and 'scores'

Show same behaviour as with events include parameter - ['scores']
same as [] and ['all_periods'] same as ['all_periods', 'scores'].

jemorriso committed:
Update tests for moneyline, total and spread

Add tests for lines from all sports.

jemorriso committed:
Update test for Rundown.schedule

jemorriso committed:
Add validator for 'Not Published' in lines

None is more semantically meaningful than 0.0001

jemorriso committed:
Handle empty / error returns in Rundown methods

jemorriso committed:
Add bad date inputs test

jemorriso committed:
Allow case-insensitive sport names

jemorriso committed:
Use Team instead of TeamNormalized

Team renamed to BaseTeam
TeamNormalized renamed to Team

jemorriso committed:
Add more tests for Rundown.schedule

jemorriso committed:
Add test for Rundown.event

jemorriso committed:
Update 'Not Published' validator

Add validation for more bad decimal numbers.

jemorriso committed:
Add use_sportsbook_names validator

Instead of having dictionary from sportsbook ID to lines, use sportsbook
name as key, which is easier for the user. The ID is still available in
the SportsbookLines / SportsbookLinePeriods attribute.

jemorriso committed:
Add static sportsbook file

Used by sportsbook validator in Event class in order to use sportsbook name
instead of ID.

jemorriso committed:
Make line class types stricter

Use make_none_if_not_published validator in order to format attributes.

jemorriso committed:
Add default empty list on optional attributes

jemorriso committed:
Add test for bad input to Rundown.events_delta

jemorriso committed:
Add missing comments on resources and functions

jemorriso committed:
Small edits

jemorriso commented:
### Commits since pull request

Addressing previous comments, adding many test cases to verify both API and client behaviour, and adding features:
- Add comments to all resources, add module level comments
- Use 'Base' instead of 'Auth' for Rundown config classes (5f88a79)
- Use `sport` instead of `sport_id` as parameter to methods, allowing user to pass in sport names such as 'NFL' or 'NHL'
  - Use static data in rundown.static.sports module (7984dbb)
- Rename `events_by_date`, `teams_by_sport`, `schedule_by_sport` to `events`, `teams`, and `schedule` respectively, because their behaviour is implied by method arguments (144653f)
- Add `Rundown._json` attribute so that the user can see the raw response of the most recent request to the API (4941d8c)
- Add `utc_shift_to_tz` to rundown.utils module, allowing user to use offset to override `Rundown.timezone` attribute (a8a1370)
- Add rundown.static.sportsbook module, `use_sportsbook_names` validator to rundown.resources.validators, allowing use of sportsbook name instead of ID as keys in the `Event` resource (2921f92, 63751d2)
- Add `make_none_if_not_published` validator so that lines that aren't available use `None` instead of 0.0001 (6769063)
- Change `Team` to  `TeamNormalized` (73d5739)

fjordan commented on rundown/resources/event.py:
```diff
@@ -0,0 +1,95 @@
+    rotation_number_home: Optional[int] = None
+    # 'score' may not be populated for games >100 days in advance.
+    score: Optional[Score] = None
+    # 'teams' not populated for games 3 days in advance.
```
Right, `teams` are only available once odds are published.

fjordan commented on rundown/resources/event.py:
```diff
@@ -0,0 +1,95 @@
+    # Rotation numbers not popuated for games 3 days in advance.
+    rotation_number_away: Optional[int] = None
+    rotation_number_home: Optional[int] = None
+    # 'score' may not be populated for games >100 days in advance.
```
Yes, we try to have score/schedule information for games around 120 days in advance.

fjordan commented on rundown/static/sports.py:
```diff
@@ -0,0 +1,43 @@
+"""Module for static sports information used by Rundown."""
```
what's the purpose of this being static? We're in the process of adding more sports today (EPL) and will definitely be adding more down the road.

jemorriso replied:
I used this static file to allow the user to pass the name of the sport as an alternative to the sport's ID to many of the methods in the `Rundown` class. Just as a convenience for the user so they can avoid remembering sport IDs.

jemorriso replied:
Using the sport name is optional - the sport parameter supports both ID and name.

fjordan replied:
Can't we make a request to the sports endpoint and use the `sport_name` to build this list dynamically instead?

fjordan commented on rundown/static/sportsbooks.py:
```diff
@@ -0,0 +1,102 @@
+"""Module for static sportsbook information used by Rundown."""
```
same to be said about sportsbooks. These will often change and more will be added.

jemorriso replied:
This one is used in `rundown.resources.event.Event` in the `use_sportsbook_names` validator. Basically the returned `Event` object has sportsbook names instead of IDs as keys for the `lines` and `line_periods` dictionaries. Again the reasoning for this is so the user can avoid trying to remember sportsbook IDs. 

The validator method would need to be changed to allow IDs as keys in case the sportsbook name isn't present in the static list.

jemorriso replied:
Right now if a new sportsbook is added that isn't in the static list, the validator will raise a ValidationError.

fjordan replied:
Same with the above - can we build this list dynamically?

jemorriso committed:
Update use_sportsbook_names validator

Use sportsbook ID as fallback key.

jemorriso committed:
Add static module, static files

Static module holds functions for building sports / sportsbook dicts.
sports.yaml and sportsbooks.yaml store cached data from server.

jemorriso committed:
Add yaml handling

jemorriso committed:
Add refresh sports / sportsbooks functionality

User can opt-in to automatically refreshing sports / sportsbook data
from server, which writes the data to sports.yaml and sportsbooks.yaml.

refresh_sports() and refresh_sportsbooks() methods added so user can
update the cached data at any time.

jemorriso committed:
Temporarily omit test

jemorriso committed:
Fix case bug - user lowercase for keys

jemorriso committed:
Accidentally deleted function return value
